metric-schema-java/src/main/java/com/palantir/metric/schema/UtilityGenerator.java

```diff
@@ -92,30 +109,196 @@ static JavaFile generateUtilityClass(
-                builder.addMethods(
-                        generateSimpleMetricFactory(namespace, metricName, libraryName, definition, visibility));
+                builder.addMethods(generateSimpleMetricFactory(
+                        namespace, metricName, libraryName, metrics, definition, visibility));
             } else {
-                generateMetricFactoryBuilder(namespace, metricName, libraryName, definition, builder, visibility);
+                generateMetricFactoryBuilder(
+                        namespace, metricName, libraryName, definition, metrics, builder, visibility);
             }
         });
 
-        builder.addMethod(generateToString(className));
+        builder.addMethod(generateToString(metrics, className));
 
         return JavaFile.builder(className.packageName(), builder.build())
                 .skipJavaLangImports(true)
                 .indent("    ")
                 .build();
     }
 
+    private static MethodSpec generateSimpleFactory(ImplementationVisibility visibility, ClassName className) {
+        return MethodSpec.methodBuilder(ReservedNames.FACTORY_METHOD)
+                .addModifiers(visibility.apply(Modifier.STATIC))
+                .addParameter(TaggedMetricRegistry.class, ReservedNames.REGISTRY_NAME)
+                .addStatement(
+                        "return new $T($T.checkNotNull(registry, $S))",
+                        className,
+                        Preconditions.class,
+                        TaggedMetricRegistry.class.getSimpleName())
+                .returns(className)
+                .build();
+    }
+
+    /** Produce a private staged builder, which implements public interfaces. */
+    private static void generateFactoryBuilder(
+            String name,
+            ClassName className,
+            MetricNamespace metricNamespace,
+            TypeSpec.Builder outerBuilder,
+            ImplementationVisibility visibility1) {
+        BuilderStage metricRegistryStage = BuilderStage.builder()
+                .name(ReservedNames.REGISTRY_NAME)
+                .sanitizedName(ReservedNames.REGISTRY_NAME)
+                .className(ClassName.get(TaggedMetricRegistry.class))
+                .build();
+        List<BuilderStage> tagStages = metricNamespace.getTags().stream()
+                .filter(UtilityGenerator::tagDefinitionRequiresParam)
+                .map(tagDef -> BuilderStage.builder()
+                        .name(tagDef.getName())
+                        .sanitizedName(Custodian.sanitizeName(tagDef.getName()))
+                        .className(getTagClassName(name, tagDef))
+                        .build())
+                .collect(Collectors.toList());
+        CodeBlock.Builder buildBlock = CodeBlock.builder();
+        buildBlock.add("new $T($L", className, metricRegistryStage.sanitizedName());
+        tagStages.forEach(stage -> buildBlock.add(", $L", stage.sanitizedName()));
+        buildBlock.add(");");
```

Comment:
Just braindumping, no action required unless you want to: javapoet provides codeblock stream collectors to make this sort of thing flow more naturaly, something along the lines of:

```java
buildBlock.addStatement("new $T($L)", className, Stream.concat...map().collect(CodeBlock....);
```

Reply:
👨‍🍳

Reply:
Almost there, can't figure out toString

Reply:
Done